csrc/fusion_segmenter.cpp

```diff
@@ -449,11 +450,39 @@ SegmentedFusion::SegmentedFusion(std::unique_ptr<Fusion> fusion)
     : segmented_fusion_name_{segmentedFusionName()},
       impl_(this),
       complete_fusion_(std::move(fusion)),
-      initial_vals_size_{0},
-      initial_exprs_size_{0} {
+      initial_vals_size_{complete_fusion_->vals().size()},
+      initial_exprs_size_{complete_fusion_->unordered_exprs().size()} {
   annotateFP16IntermediateTensors();
 }
 
+namespace {
+
+//! Check if all values and expressions in SegmentedGroup are compatible with
+//! initial complete fusion
+bool isCompatibleSegmentedGroup(
+    SegmentedGroup* sg,
+    const std::unordered_map<Val*, int64_t>& vals_to_id_map,
+    const std::unordered_map<Expr*, int64_t>& exprs_to_id_map,
+    int64_t initial_vals_size,
+    int64_t initial_exprs_size) {
+  auto check_value = [&](Val* v) {
+    return vals_to_id_map.at(v) < initial_vals_size;
+  };
+  auto check_expr = [&](Expr* e) {
+    return exprs_to_id_map.at(e) < initial_exprs_size;
+  };
+  bool all_compatible_inputs =
+      std::all_of(sg->inputs().begin(), sg->inputs().end(), check_value);
+  bool all_compatible_outputs =
+      std::all_of(sg->outputs().begin(), sg->outputs().end(), check_value);
```

Comment:
This is assuming all intermediate `Val*` would have an entry that's smaller than the largest outputs. I guess we have incremental update so that has to be safe for now.

I also lean towards checking all vals leading to output just to be "future-proof"... not a strong opinion since it has no impact for today.

Reply:
The `input_vals` and `output_vals` are the only `Val*` in `SegmentedGroup`, so there aren't any intermediates to check for serialization. This function checks if we can serialize correctly. It doesn't guarantee that the segmented group is correct.